src/lib/ProductExporter.ts

```diff
@@ -503,27 +510,68 @@ export class ProductExporter {
         return chargeToSave;
     }
 
-    private async retrieveBundleElements(connection: Connection, productList: Set<string>){
-        let bundleElements = await Queries.queryBundleElements(connection, productList);  
+    private async getBundleElementsByBundleTechIds(connection: Connection, bundleTechIds: Set<string>): Promise<Array<any>>{
+        let bundleElements = await Queries.queryBundleElementsByBundleTechIds(connection, bundleTechIds);  
         this.checkTechIds(bundleElements);
 
+        return this.prepareBundleElementsStructure(connection, bundleElements);
+    }
+
+    private async prepareBundleElementsStructure(connection: Connection, bundleElementsFromQuery: Array<any>): Promise<Array<any>>{
         let bundleElementOptions = await Queries.queryBundleElementOptions(
             connection, 
-            new Set(bundleElements.map(bundleElement => bundleElement['enxCPQ__TECH_External_Id__c']))
+            new Set(bundleElementsFromQuery.map(bundleElement => bundleElement['enxCPQ__TECH_External_Id__c']))
         );
         this.checkTechIds(bundleElementOptions);
-
-        for(let bundleElement of bundleElements){
-            const bundleElementOptionsToSave = bundleElementOptions.filter(option => (
+        
+        const bundleElementsStructuredForSave = bundleElementsFromQuery.map(bundleElement => {
+            const options = bundleElementOptions.filter(option => (
                 option['enxCPQ__Bundle_Element__r'] && option['enxCPQ__Bundle_Element__r']['enxCPQ__TECH_External_Id__c'] === bundleElement['enxCPQ__TECH_External_Id__c']
             ));
-            const elementToSave = {
+            return {
                 root: bundleElement,
-                bundleElementOptions: bundleElementOptionsToSave
+                bundleElementOptions: options
             }
+        });
 
-            const filePath = '/bundleElements/' + Util.sanitizeFileName(bundleElement['Name']) + '_' + bundleElement['enxCPQ__TECH_External_Id__c'] + '.json';
-            Util.writeFile(filePath, elementToSave);
+        return bundleElementsStructuredForSave;
+    }
+
+
+    private async retrieveBundleElements(connection: Connection, productList: Set<string>): Promise<void> {
+        const bundleElementsFromRootProductsQuery = await Queries.queryBundleElements(connection, productList);
+        let bundleElementsToSave = await this.prepareBundleElementsStructure(connection, bundleElementsFromRootProductsQuery);
+        let shouldQueryAdditionalBundleElements = true;
+
+        while(shouldQueryAdditionalBundleElements){
+            const additionalProductsTechIdsToQueryBy = Util.convert2DTo1DArray(bundleElementsToSave.map(bundleElement => {
+                const bundleElementOptionsProductsTechIds = bundleElement.bundleElementOptions.map(option => {
+                    if(option['enxCPQ__Product__r']){
```

Comment:
could be early return 
and a little bit of empty lines would improve readability

Reply:
Oh! I have better idea! Not early return. Filter them before map!